### ISSUE TITLE
BackupPC does not daemonize

### ISSUE THREAD

**Issue by sigmoidal** [backuppc/backuppc#64]
I am trying to manually  execute as a non-root backuppc user: /usr/bin/BackupPC -d but it seems that it fails right after: https://github.com/backuppc/backuppc/blob/master/bin/BackupPC#L312

If I run backuppc without the "-d" option, it stays active in the foreground. 

There is nothing being logged. However, I checked that $pid does get assigned a normal process ID, but it dies right after L312. 

I run BackupPC with warnings turned on to see if there is anything interesting:
```
"my" variable $lockFd masks earlier declaration in same scope at /usr/lib/BackupPC/Storage/Text.pm line 474.
"my" variable $locked masks earlier declaration in same scope at /usr/lib/BackupPC/Storage/Text.pm line 474.
Statement unlikely to be reached at /usr/lib/BackupPC/Lib.pm line 1166.
        (Maybe you meant system() when you said exec()?)
Statement unlikely to be reached at /usr/lib/BackupPC/Lib.pm line 1231.
        (Maybe you meant system() when you said exec()?)
Statement unlikely to be reached at /usr/bin/BackupPC line 601.
        (Maybe you meant system() when you said exec()?)
Statement unlikely to be reached at /usr/bin/BackupPC line 773.
        (Maybe you meant system() when you said exec()?)
Possible attempt to separate words with commas at /usr/bin/BackupPC line 1350.
Name "BackupPC::Storage::Text::LOCK" used only once: possible typo at /usr/lib/BackupPC/Storage/Text.pm line 482.
Use of uninitialized value $topDir in string eq at /usr/lib/BackupPC/Lib.pm line 70.
Use of uninitialized value $installDir in string eq at /usr/lib/BackupPC/Lib.pm line 71.
Use of uninitialized value $confDir in string eq at /usr/lib/BackupPC/Lib.pm line 79.
Use of uninitialized value $host in string ne at /usr/lib/BackupPC/Lib.pm line 348.
binmode() on closed filehandle CHILD at /usr/lib/BackupPC/Lib.pm line 1214.
Use of uninitialized value in string ne at /usr/bin/BackupPC line 271.
```

I am on gentoo x64 and using perl v5.24.1. I wrote my own ebuilds (maybe that contributes to the problem?) and I'm using BackupPC:XS-v0.52 w/ BackupPC v4.0.0

Any ideas what may be wrong? or anything interesting to try?

thanks in advance.

**Comment by craigbarratt** [comment]
I'm running perl 5.22.1.  I'm not sure if this issue is related to 5.24.1, or something else with your configuration.

Can you run the perl debugger and see what happens:
`perl -d /usr/local/BackupPC/bin/BackupPC
t 0
r
`

The "t 0" command turns on tracing at the current stack level (ie: subroutines are not traced).  It will print each line of code as it is executed, so you should be able to see where it fails/stops/exits.

**Comment by sigmoidal** [comment]
It outputs:

```
main::Main_Initialize(/usr/bin/BackupPC:306):
306:        if ( $opts{d} ) {
main::Main_Initialize(/usr/bin/BackupPC:311):
311:            my $pid;
main::Main_Initialize(/usr/bin/BackupPC:312):
312:            defined($pid = fork) or die("Can't fork: $!");
main::Main_Initialize(/usr/bin/BackupPC:313):
313:            exit if ( $pid );   # parent exits
Debugged program terminated.  Use q to quit or R to restart,
use o inhibit_exit to avoid stopping after program termination,
h q, h R or h o to get additional info.
```

like before really. Note however, that I cannot use the latest git tree, I use the released 4.0.0.
I opened another issue regarding this. I have to work out how things changed and what I have to do to get it working again.

any ideas welcome.

**Comment by sigmoidal** [comment]
I upgraded from 4.0.0. to git master, but it still fails. I used to run a 3.x backuppc that worked fine with perl 5.24.1, so I can't really tell why this new version isn't behaving.

**Comment by craigbarratt** [comment]
Unfortunately the debugger doesn't track the forked child, and the parent
simply exits, so we didn't learn anything from that.

The daemonize code hasn't changed between 3.x and 4.0.0.

What happens when you start BackupPC without the -d option (so it doesn't
daemonize)?

What happens when you run the debugger trace on BackupPC without the -d
option (so it doesn't daemonize)?

Craig

**Comment by sigmoidal** [comment]
That's right yes. I installed from cpan Devel:Trace:Fork and I run BackupPC with and without the -d option:
```
perl -d:Trace::Fork /usr/bin/BackupPC -d
perl -d:Trace::Fork /usr/bin/BackupPC
```
Unfortunately, GitHub does not let me attach the files to this post (for whatever reason it doesn't accept the zip file!), so I put it here temporarily:

https://mega.nz/#!Q951DahI!Ndgfx_T9YW8QhiG_8qlZvVIZ8fK3Xt4BtR-41u-XcwE

if you want to take a look. I couldn't see anything obvious, but I don't know backuppc that well.
If you have any ideas, let me know I can try.

thanks.

**Comment by craigbarratt** [comment]
Thanks for the links.  But I can't unzip the archive (error is: need PK
compat. v6.3 (can do v4.5)).

The logs are quite small.  Can you just email me a tar ball?

Craig

**Comment by craigbarratt** [comment]
I installed 7zip and was able to unpack the archive.  Looking at the traces now...

**Comment by craigbarratt** [comment]
Thanks for the traces.  Using Trace is a great idea!

The non-daemon run (fg) is really informative.  After 5 seconds, BackupPC receives a signal which causes it to exit.  The BackupPC server LOG file should contain the name of the signal.  Can you look there and tell me what signal it is?

**Comment by sigmoidal** [comment]
Craig, I think my first post was not clear enough and I confused you, sorry. 

When running BackupPC without the -d option there is no problem. The process stays active in the foreground as expected. The fg log shows me exiting the active process with Ctrl+C after a few seconds (otherwise I'd have to wait for hours till next wake up). 

I rerun it and waited for a while (until next wake up). I attach the logs, but I'm not sure how useful they are. The one log is from the backuppc LOG file (maybe useful to weed out some warnings), while the other one is from DTrace. The process runs fine, I don't see anything strange.

I wonder what is wrong.

[backuppc_4.x_fork_child_dies_v2.tar.gz](https://github.com/backuppc/backuppc/files/846822/backuppc_4.x_fork_child_dies_v2.tar.gz)

**Comment by craigbarratt** [comment]
Ah, thanks for clarifying that.  Ok, so looking at the -d log, there is no mention of the child process that should be created by the fork() call.  That's really strange - it's as though fork() is failing on your system.  But fork() is so fundamental I can't imagine it can fail (eg, in the non-daemon case, fork() is used to run every backup and sub-command).

What happens when you run a really simple test program, for example:

    if (fork() == 0) {
        print "This is the child.\n";
        exit 0;
    }
    print "This is the parent.\n";

and run it with:

    perl -d:Trace::Fork forkTest

The output I get is:

    >  /usr/local/bin/perl forkTest
    > 1489703044 1489703044.85963
    >>    0.860631:31719:0:fork:3: if (fork() == 0) {
    >>    0.861022:31719:0:fork:7: print "This is the parent.\n";
    >>    0.861291:31719: -------------- END OF PROCESS ------------------
    >>    0.861490:31720: ----------------- NEW PROCESS -----------------
    >>    0.861490:31720:0:fork:4:     print "This is the child.\n";
    >>    0.862236:31720:0:fork:5:     exit 0;
    >>    0.862341:31720: -------------- END OF PROCESS ------------------

You can see the parent exits (pid 31719), and the trace correctly shows the new child process is running (31720).

You could then modify the script to make it use more of the exact code BackupPC uses, eg:

    #!/usr/bin/perl

    use POSIX qw(setsid);

    my $pid;
    defined($pid = fork) or die("Can't fork: $!");
    exit if ( $pid );   # 1st parent exits

    print "This is the 1st child.\n";

    POSIX::setsid();
    defined($pid = fork) or die("Can't fork: $!");
    exit if $pid;   # 2nd parent exits

    print "This is the 2nd child.\n";

which produces this output on my system:

    >  /usr/local/bin/perl forkTest2
    > 1489703668 1489703668.15138
    >>    0.156349:31767:0:forkTest2:5: my $pid;
    >>    0.156375:31767:0:forkTest2:7: defined($pid = fork) or die("Can't fork: $!");
    >>    0.156692:31767:0:forkTest2:8: exit if ( $pid );   # 1st parent exits
    >>    0.156818:31767: -------------- END OF PROCESS ------------------
    >>    0.157089:31768: ----------------- NEW PROCESS -----------------
    >>    0.157089:31768:0:forkTest2:8: exit if ( $pid );   # 1st parent exits
    >>    0.157695:31768:0:forkTest2:10: print "This is the 1st child.\n";
    >>    0.157731:31768:0:forkTest2:12: POSIX::setsid();
    >>    0.157834:31768:0:forkTest2:13: defined($pid = fork) or die("Can't fork: $!");
    >>    0.158151:31768:0:forkTest2:14: exit if $pid;   # 2nd parent exits
    >>    0.158381:31768: -------------- END OF PROCESS ------------------
    >>    0.158524:31769: ----------------- NEW PROCESS -----------------
    >>    0.158524:31769:0:forkTest2:14: exit if $pid;   # 2nd parent exits
    >>    0.159972:31769:0:forkTest2:16: print "This is the 2nd child.\n";
    >>    0.160100:31769: -------------- END OF PROCESS ------------------

Craig

**Comment by sigmoidal** [comment]
Craig,

Other daemons seem to work fine and this is a very well behaving server system in general. I think fork has no issues overall, unless my perl is done, but I would be having other issues probably.. I hope we find it.

I made the two files (forkTest and forkTest2), exactly as you suggest.
I first ran forkTest:
```
$ perl -d:Trace::Fork forkTest
This is the parent.
This is the child.
Devel::Trace::Fork output in <./DTFork.15073>


$ cat DTFork.15073
>  /usr/bin/perl forkTest 
> 1489704538 1489704538.63056
>>    0.632124:15073:0:forkTest:1: if (fork() == 0) {
>>    0.632412:15073:0:forkTest:5: print "This is the parent.\n";
>>    0.632516:15074: ----------------- NEW PROCESS -----------------
>>    0.632597:15073: -------------- END OF PROCESS ------------------
>>    0.632516:15074:0:forkTest:2:     print "This is the child.\n";
>>    0.632927:15074:0:forkTest:3:     exit 0;
>>    0.633001:15074: -------------- END OF PROCESS ------------------
```

I then ran forkTest2:

```
$ perl -d:Trace::Fork forkTest2

This is the 1st child.
Devel::Trace::Fork output in <./DTFork.15180>
Devel::Trace::Fork: possibly appending trace data from 15182 after main process 15180 exited.
 at /usr/lib64/perl5/vendor_perl/5.24.1/Devel/Trace/Fork.pm line 80.
        Devel::Trace::Fork::flush() called at /usr/lib64/perl5/vendor_perl/5.24.1/Devel/Trace/Fork.pm line 49
        DB::DB called at forkTest2 line 13
Devel::Trace::Fork: possibly appending trace data from 15181 after main process 15180 exited.
 at /usr/lib64/perl5/vendor_perl/5.24.1/Devel/Trace/Fork.pm line 80.
        Devel::Trace::Fork::flush() called at /usr/lib64/perl5/vendor_perl/5.24.1/Devel/Trace/Fork.pm line 110
        Devel::Trace::Fork::END() called at forkTest2 line 0
        eval {...} called at forkTest2 line 0
This is the 2nd child.
Devel::Trace::Fork output in <./DTFork.15180>

$ cat DTFork.15180
>  /usr/bin/perl forkTest2 
> 1489704729 1489704729.78137
>>    0.790996:15180:0:forkTest2:5: my $pid;
>>    0.791032:15180:0:forkTest2:6: defined($pid = fork) or die("Can't fork: $!");
>>    0.791352:15180:0:forkTest2:7: exit if ( $pid );   # 1st parent exits
>>    0.791443:15181: ----------------- NEW PROCESS -----------------
>>    0.791443:15181:0:forkTest2:7: exit if ( $pid );   # 1st parent exits
>>    0.791498:15180: -------------- END OF PROCESS ------------------
>>    0.791783:15181:0:forkTest2:9: print "This is the 1st child.\n";
>>    0.791826:15181:0:forkTest2:11: POSIX::setsid();
>>    0.791894:15181:0:forkTest2:12: defined($pid = fork) or die("Can't fork: $!");
>>    0.792330:15181:0:forkTest2:13: exit if $pid;   # 2nd parent exits
>>    0.792356:15182: ----------------- NEW PROCESS -----------------
>>    0.792356:15182:0:forkTest2:13: exit if $pid;   # 2nd parent exits
>>    0.792515:15181: -------------- END OF PROCESS ------------------
>>    0.793283:15182:0:forkTest2:15: print "This is the 2nd child.\n";
>>    0.793401:15182: -------------- END OF PROCESS ------------------


I don't see anything unusual. Do you?

Thanks for helping.

**Comment by craigbarratt** [comment]
Yes, that works as expected.  I can't think of why BackupPC fails to fork (or why the tracing fails to trace the child process).

In bin/BackupPC, can you add a print just after the 2nd fork() (before the chdir ("/")) and see if it prints anything? 

One way to debug further is to keep adding lines of code from the beginning of BackupPC to your test script and see at what point fork() starts failing (ie: no trace output from the child process).  BackupPC calls Main_Initialize() relatively early in the startup and it doesn't do that much before it tries to daemonize.  You'll need to bisect to find what code causes the failure.

**Comment by sigmoidal** [comment]
Craig, 

The problem is related to a call made to getpwnam (also maybe through perl libs) and /etc/nsswitch.conf. One of my lines is:

```
shadow:     files    ldap
```

if I remove the ldap part of that line (i have other lines that have it and there seems to be no problem),  BPC daemonizes as expected.

I have not fully understood why this happens yet, there are barely any logs. I was accidentally lead to this by looking at connections happening to slapd and I hypothesized that this may be related, so I started trying to track down the problem.

I don't know why daemonizing and these are related. It could be that this is a problem beyond BPC, but I'd sure like to get a better understanding, so if you have any thoughts let me know.

I will write a PoC without BPC to see if I can reproduce it soon (I spend so many days debugging I need a break), then we can point the finger to either Perl libs, system config or BPC. :)

thanks.

**Comment by sigmoidal** [comment]
I went on and made a PoC using the perl getpwname(), without BPC and still using Perl 5.24.1 on Gentoo amd64. The problem persisted, which definately rules out a problem in BPC.

For what its worth it, I will describe the fix (and close the issue), because there may be other people experiencing the same problem, in BPC or another context.

I was using for ldap integration on my system the gentoo packages sys-auth/nss_ldap and sys-auth/pam_ldap (http://www.padl.com/OSS). Turns out that these pieces of software are dated and they can be replaced by sys-auth/nss-pam-ldapd (http://arthurdejong.org/nss-pam-ldapd).

Once I replaced the files, configured and run nslcd, I used the following proof of concept to verify processes can be daemonized:
```
print join(",", getpwnam("backuppc"));

if (fork() == 0) {
    print "This is the child: " . $$ . "\n";
    print "If you see this, fork is working correctly.\n";
    exit 0;
}
print "This is the parent: " . $$ . "\n";
```

This example did not execute the code inside the if statement at all when the problem was present.
Anyway, I hope this helps someone at some point, although it must be clarified it is not BPC related.

thanks for the interest and help.
